### PR TITLE
PHAL

### PR DESCRIPTION
**Pull request by JarbasAl**
this goes together with https://github.com/OpenVoiceOS/ovos_utils/pull/11

what used to be the enclosure template in ovos utils is now the PHAL plugin

### CODE REVIEW

File: ovos_plugin_manager/templates/phal.py [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+
+    def run(self):
+        ''' start enclosure '''
+        self._running = True
+        while self._running:
+            time.sleep(1)
+
+    # Audio Events
+    def on_record_begin(self, message=None):
+        ''' listening started '''
```
Comment (NeonDaniel) [inline ovos_plugin_manager/templates/phal.py:129]: Consider updating `'''` to `"""` in docstrings for consistency

File: ovos_plugin_manager/templates/phal.py [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+        Typically this would be represented by the eyes being 'open'
+        and the mouth reset to its default (smile or blank).
+        """
+        pass
+
+    # System Events
+    def on_no_internet(self, message=None):
+        """
+
+        Args:
```
Comment (NeonDaniel) [inline ovos_plugin_manager/templates/phal.py:169]: This should either be added to all handlers or removed from this one

File: ovos_plugin_manager/templates/phal.py [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+        pass
+
+    def on_system_unmute(self, message=None):
+        """Unmute (turn on) the system speaker."""
+        pass
+
+    def on_system_blink(self, message=None):
+        """The 'eyes' should blink the given number of times.
+        Args:
+            times (int): number of times to blink
```
Comment (NeonDaniel) [inline ovos_plugin_manager/templates/phal.py:189]: args don't match signature

File: ovos_plugin_manager/templates/phal.py [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+    def on_eyes_on(self, message=None):
+        """Illuminate or show the eyes."""
+        pass
+
+    def on_eyes_off(self, message=None):
+        """Turn off or hide the eyes."""
+        pass
+
+    def on_eyes_fill(self, message=None):
+        pass
```
Comment (NeonDaniel) [inline ovos_plugin_manager/templates/phal.py:203]: Docstring missing

File: ovos_plugin_manager/templates/phal.py [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+
+    def on_eyes_blink(self, message=None):
+        """Make the eyes blink
+        Args:
+            side (str): 'r', 'l', or 'b' for 'right', 'left' or 'both'
```
Comment (NeonDaniel) [inline ovos_plugin_manager/templates/phal.py:208]: args don't match signature

File: ovos_plugin_manager/templates/phal.py [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+        pass
+
+    def on_eyes_narrow(self, message=None):
+        """Make the eyes look narrow, like a squint"""
+        pass
+
+    def on_eyes_look(self, message=None):
+        """Make the eyes look to the given side
+        Args:
+            side (str): 'r' for right
```
Comment (NeonDaniel) [inline ovos_plugin_manager/templates/phal.py:219]: args don't match signature

File: ovos_plugin_manager/templates/phal.py [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+                        'l' for left
+                        'u' for up
+                        'd' for down
+                        'c' for crossed
+        """
+        pass
+
+    def on_eyes_color(self, message=None):
+        """Change the eye color to the given RGB color
+        Args:
```
Comment (NeonDaniel) [inline ovos_plugin_manager/templates/phal.py:229]: args don't match signature

File: ovos_plugin_manager/templates/phal.py [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+            r (int): 0-255, red value
+            g (int): 0-255, green value
+            b (int): 0-255, blue value
+        """
+        pass
+
+    def on_eyes_brightness(self, message=None):
+        """Set the brightness of the eyes in the display.
+        Args:
```
Comment (NeonDaniel) [inline ovos_plugin_manager/templates/phal.py:238]: args don't match signature

File: ovos_plugin_manager/templates/phal.py [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+        pass
+
+    def on_eyes_reset(self, message=None):
+        """Restore the eyes to their default (ready) state."""
+        pass
+
+    def on_eyes_timed_spin(self, message=None):
+        """Make the eyes 'roll' for the given time.
+        Args:
+            length (int): duration in milliseconds of roll, None = forever
```
Comment (NeonDaniel) [inline ovos_plugin_manager/templates/phal.py:250]: args don't match signature

File: ovos_plugin_manager/templates/phal.py [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+        """
+        pass
+
+    def on_eyes_volume(self, message=None):
+        """Indicate the volume using the eyes
+        Args:
+            volume (int): 0 to 11
```
Comment (NeonDaniel) [inline ovos_plugin_manager/templates/phal.py:257]: args don't match signature

File: ovos_plugin_manager/templates/phal.py [hunk 11]
```diff
@@ -0,0 +1,408 @@
...
+        """
+        pass
+
+    def on_eyes_spin(self, message=None):
+        """
+        Args:
```
Comment (NeonDaniel) [inline ovos_plugin_manager/templates/phal.py:263]: Update docstring

File: ovos_plugin_manager/templates/phal.py [hunk 12]
```diff
@@ -0,0 +1,408 @@
...
+        """
+        pass
+
+    def on_eyes_set_pixel(self, message=None):
+        """
+        Args:
```
Comment (NeonDaniel) [inline ovos_plugin_manager/templates/phal.py:269]: Update docstring

File: ovos_plugin_manager/templates/phal.py [hunk 13]
```diff
@@ -0,0 +1,408 @@
...
+
+    def _on_mouth_smile(self, message=None):
+        """Show a 'smile' image or animation"""
+        if self.mouth_events_active:
+            self.on_smile(message)
+
+    def _on_mouth_viseme(self, message=None):
+        """Display a viseme mouth shape for synched speech
+        Args:
+            code (int):  0 = shape for sounds like 'y' or 'aa'
```
Comment (NeonDaniel) [inline ovos_plugin_manager/templates/phal.py:302]: args don't match signature

File: ovos_plugin_manager/templates/phal.py [hunk 14]
```diff
@@ -0,0 +1,408 @@
...
+                         5 = shape for sounds like 'f' or 'v'
+                         6 = shape for sounds like 'oy' or 'ao'
+        """
+        if self.mouth_events_active:
+            self.on_viseme(message)
+
+    def _on_mouth_text(self, message=None):
+        """Display text (scrolling as needed)
+        Args:
+            text (str): text string to display
```
Comment (NeonDaniel) [inline ovos_plugin_manager/templates/phal.py:316]: args don't match signature

File: ovos_plugin_manager/templates/phal.py [hunk 15]
```diff
@@ -0,0 +1,408 @@
...
+        """
+        if self.mouth_events_active:
+            self.on_text(message)
+
+    def _on_mouth_display(self, message=None):
+        if self.mouth_events_active:
```
Comment (NeonDaniel) [inline ovos_plugin_manager/templates/phal.py:322]: missing docstring

File: ovos_plugin_manager/templates/phal.py [hunk 16]
```diff
@@ -0,0 +1,408 @@
...
+        pass
+
+    def on_smile(self, message=None):
+        """Show a 'smile' image or animation"""
+        pass
+
+    def on_viseme(self, message=None):
+        """Display a viseme mouth shape for synched speech
+        Args:
+            code (int):  0 = shape for sounds like 'y' or 'aa'
```
Comment (NeonDaniel) [inline ovos_plugin_manager/templates/phal.py:349]: args don't match signature

File: ovos_plugin_manager/templates/phal.py [hunk 17]
```diff
@@ -0,0 +1,408 @@
...
+            text (str): text string to display
+        """
+        pass
+
+    def on_display(self, message=None):
+        """Display images on faceplate. Currently supports images up to 16x8,
+           or half the face. You can use the 'x' parameter to cover the other
+           half of the faceplate.
+        Args:
+            img_code (str): text string that encodes a black and white image
```
Comment (NeonDaniel) [inline ovos_plugin_manager/templates/phal.py:371]: args don't match signature

File: ovos_plugin_manager/templates/phal.py [hunk 18]
```diff
@@ -0,0 +1,408 @@
...
+                            Useful if you'd like to display muliple images
+                            on the faceplate at once.
+        """
+        pass
+
+    def on_weather_display(self, message=None):
+        """Show a the temperature and a weather icon
+
+        Args:
+            img_code (char): one of the following icon codes
```
Comment (NeonDaniel) [inline ovos_plugin_manager/templates/phal.py:385]: args don't match signature

**Review by NeonDaniel — CHANGES_REQUESTED** [review]
Noted missing/incorrect documentation

**Comment by JarbasAl** [comment]
this PR is pending on the actual release of PHAL and may still change significantly, marking as draft

File: ovos_plugin_manager/templates/phal.py (hunk 3 above)
Comment (JarbasAl) [reply ovos_plugin_manager/templates/phal.py:189]: in all these args they are documenting what comes inside the message.data , they are not final and were copied over from original methods (usually from mark1 code)

as you can probably tell i copied over almost every defined event handler for enclosure in mycroft-core

**Comment by JarbasAl** [comment]
this goes together with https://github.com/OpenVoiceOS/ovos_utils/pull/11

what used to be the enclosure template in ovos utils is now the PHAL plugin

File: ovos_plugin_manager/templates/phal.py (hunk 3 above)
Comment (NeonDaniel) [reply ovos_plugin_manager/templates/phal.py:189]: Yeah; figured it was copy/paste.. Maybe we should define some standard way to document expected data/context in these docstrings?

**Comment by JarbasAl** [comment]
early merging to allow PHAL release 0.0.1

will refactor later

**Pull request merged**